### PR TITLE
Compatibility with PlaceholderAPI

### PR DESCRIPTION
**Pull request by Phoetrix**
Ok so in my test, everything work correctly, I hope that I forgot nothing. ^^

But you need to add PlaceholderAPI-2.8.2.jar in the library to allow my class to work. => https://www.spigotmc.org/resources/placeholderapi.6245/

I am a beginner in Java so I don't known if there is a better way to do what I do so don't hesitate to tell me the problem.
With this, users will be able to get achievement data of players in others plugins by using placeholders.

All placeholders return a string.
Here is the list of placeholders:

%aach_connections%
%aach_deaths%
%aach_arrows%
%aach_snowballs%
%aach_eggs%
%aach_fish%
%aach_treasures%
%aach_itembreaks%
%aach_eatenitems%
%aach_shear%
%aach_milk%
%aach_lavabuckets%
%aach_waterbuckets%
%aach_trades%
%aach_anvilsused%
%aach_enchantements%
%aach_beds%
%aach_maxlevel%
%aach_consumedpotions%
%aach_playedtime%
%aach_distancefoot%
%aach_distancepig%
%aach_distancehorse%
%aach_distanceminecart%
%aach_distanceboat%
%aach_distancegliding%
%aach_distancellama%
%aach_itemdrops%
%aach_itempickups%
%aach_hoeplowings%
%aach_fertilising%
%aach_taming%
%aach_brewing%
%aach_fireworks%
%aach_musicdiscs%
%aach_enderpearls%
%aach_smelting%
%aach_petmastergive%
%aach_petmasterreceive%

%aach_places_[blockname]% - example: %aach_places_dirt%
%aach_breaks_[blockname]% - example: %aach_breaks_stone%
%aach_kills_[entityname]% - example: %aach_kills_zombie%
%aach_crafts_[itemname]% - example: %aach_crafts_bread%
%aach_playercommands[command]% - example: %aach_playercommands_aach list%
%aach_breeding_[entityname]% - example: %aach_breeding_pig%

### CODE REVIEW

**Comment by Phoetrix** [comment]
Ah yes I forgot the line:
`import com.hm.achievement.listener.PlaceholderAchievementListener;`

In the main class, sorry. :(

**Comment by PyvesB** [comment]
Thanks for your contribution, I will review this tomorrow evening. You can directly modify your pull request by pushing commits to your fork! 👍

**Comment by Phoetrix** [comment]
Ok good! I used your API so if there is a simpler way in your plugin to get players data, give it to me.
This will be useful if I want to add new features later. :)

**Comment by supertassu** [comment]
Ref #268

File: src/main/java/com/hm/achievement/AdvancedAchievements.java [hunk 1]
```diff
@@ -678,6 +680,11 @@ private void registerListeners() {
 
 		playerAdvancedAchievementListener = new PlayerAdvancedAchievementListener(this);
 		pm.registerEvents(playerAdvancedAchievementListener, this);
+		
+		if (Bukkit.getPluginManager().isPluginEnabled("PlaceholderAPI"))
```
Comment (PyvesB) [inline src/main/java/com/hm/achievement/AdvancedAchievements.java]: Could this block of code please be placed next to the `BungeeTabListPlusBukkitAPI` bit further up in the class? See my comment below about it not really being a listener.

File: src/main/java/com/hm/achievement/listener/PlaceholderAchievementListener.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import com.hm.achievement.AdvancedAchievements;
+import com.hm.achievement.api.AdvancedAchievementsAPI;
+import com.hm.achievement.category.MultipleAchievements;
+import com.hm.achievement.category.NormalAchievements;
+
+import me.clip.placeholderapi.external.EZPlaceholderHook;
+
+public class PlaceholderAchievementListener extends EZPlaceholderHook
+{
+	AdvancedAchievements plugin;
```
Comment (PyvesB) [inline src/main/java/com/hm/achievement/listener/PlaceholderAchievementListener.java]: Please add the `private final` modifier to this field. We don't intend to modify it or use it from another class. ^^

File: src/main/java/com/hm/achievement/listener/PlaceholderAchievementListener.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	@Override
+	public String onPlaceholderRequest(Player p, String identifier)
+	{
+		for (NormalAchievements category : NormalAchievements.values())
+		{
+			String categoryName = category.toString();
+			
+			if (identifier.equalsIgnoreCase(categoryName))
+			{
+				return String.valueOf(api.getStatisticForNormalCategory(p.getUniqueId(), category));
```
Comment (PyvesB) [inline src/main/java/com/hm/achievement/listener/PlaceholderAchievementListener.java]: Using the API is definitely great when you're integrating from a different plugin, but you can directly use the internals of the plugin here:
`return plugin.getCacheManager().getAndIncrementStatisticAmount(category, p.getUniqueId(), 0);`

Therefore you don't actually need the `AdvancedAchievementsAPI ` object anymore and you squeeze in a little bit of performance as the API layer does some extra checks underneath the hood.

File: src/main/java/com/hm/achievement/listener/PlaceholderAchievementListener.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+		{
+			String categoryName = category.toString();
+
+			for (String section : plugin.getPluginConfig().getConfigurationSection(categoryName).getKeys(false))
+			{
+				String categoryPath = categoryName+"_"+section;
+				
+				if (identifier.equalsIgnoreCase(categoryPath))
+				{
+					return String.valueOf(api.getStatisticForMultipleCategory(p.getUniqueId(), category, section));
```
Comment (PyvesB) [inline src/main/java/com/hm/achievement/listener/PlaceholderAchievementListener.java]: For Multiple categories, the call would be:
`return pluginInstance.getCacheManager().getAndIncrementStatisticAmount(category, subcategory, p.getUniqueId(), 0);`

File: src/main/resources/plugin.yml [hunk 5]
```diff
@@ -2,7 +2,7 @@ name: AdvancedAchievements
 main: com.hm.achievement.AdvancedAchievements
 version: ${project.version}
 author: DarkPyves
-softdepend: [Vault,PetMaster,BungeeTabListPlus,Essentials]
+softdepend: [Vault,PetMaster,BungeeTabListPlus,Essentials,PlaceholderAPI]
```
Comment (PyvesB) [inline src/main/resources/plugin.yml]: Nice catch! People, including myself, often forget to add this. ^^

File: src/main/java/com/hm/achievement/listener/PlaceholderAchievementListener.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import org.bukkit.entity.Player;
+
+import com.hm.achievement.AdvancedAchievements;
+import com.hm.achievement.api.AdvancedAchievementsAPI;
+import com.hm.achievement.category.MultipleAchievements;
+import com.hm.achievement.category.NormalAchievements;
+
+import me.clip.placeholderapi.external.EZPlaceholderHook;
+
+public class PlaceholderAchievementListener extends EZPlaceholderHook
```
Comment (PyvesB) [inline src/main/java/com/hm/achievement/listener/PlaceholderAchievementListener.java]: Is this class actually a listener? I haven't thought this through, but I would be tempted to say it should be called something like `AchievementPlaceholderHook` and moved to the `utils` package.

File: src/main/java/com/hm/achievement/listener/PlaceholderAchievementListener.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+import org.bukkit.entity.Player;
+
+import com.hm.achievement.AdvancedAchievements;
+import com.hm.achievement.api.AdvancedAchievementsAPI;
+import com.hm.achievement.category.MultipleAchievements;
+import com.hm.achievement.category.NormalAchievements;
+
+import me.clip.placeholderapi.external.EZPlaceholderHook;
+
```
Comment (PyvesB) [inline src/main/java/com/hm/achievement/listener/PlaceholderAchievementListener.java]: Can you please add a Javadoc comment briefly explaining what this class does (see `AchievementCountBungeeTabListPlusVariable.java`)? Don't forget to put your own name as the author of the class for posterity! 😃

**Comment by Phoetrix** [comment]
Thank you very much for your review, I appreciate the time you spend to correct everything! You are nice. ^^
And I didn't know about importing of placeholderapi with reference in the pom. Maven is unbelievable!

Now I think that everything is okay!

**Comment by PyvesB** [comment]
Everything looks good to me now, thanks! Before I merge, can you please make the `AchievementPlaceholderHook.java` class  (and only that one) go through Eclipse's code formatter so that the style is consistent? I believe the default shortcut is `Ctrl` + `Shift` + `F`.

**Comment by Phoetrix** [comment]
Ok :)

**Pull request merged**